openid-data-classes/src/commonMain/kotlin/at/asitplus/openid/RequestParametersFromSerializer.kt

```diff
@@ -0,0 +1,117 @@
+import kotlinx.serialization.json.JsonDecoder
+import kotlinx.serialization.json.JsonEncoder
+import kotlinx.serialization.json.buildJsonObject
+import kotlinx.serialization.json.decodeFromJsonElement
+import kotlinx.serialization.json.encodeToJsonElement
+import kotlinx.serialization.json.jsonObject
+
+/**
+ * In order to de-/serialize generic types we need a kind of factory approach
+ * Because we deal with a sealed class we can use an intermediary jsonSerializer,
+ * find the correct object and the specific type of the generic type and
+ * then finalize the serialization
+ *
+ * In order to de-/serialize JwsSigned which itself is again a generic class
+ * we use the fact that we can find the class of parameters before we need to know the
+ * generic class of JwsSigned. To serialize we use [JwsSignedSerializer].
+ */
+class RequestParametersFromSerializer<T : RequestParameters>(
+    private val parameterSerializer: KSerializer<T>,
+) : KSerializer<RequestParametersFrom<T>> {
+    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("RequestParametersFromClass") {
+        element("Json", buildClassSerialDescriptor("Json") {
+            element<String>("jsonString")
+            element("parameters", parameterSerializer.descriptor)
+        })
+        element("JwsSigned", buildClassSerialDescriptor("JwsSigned") {
+            element("jwsSigned", JwsSignedSerializer.descriptor)
+            element("parameters", parameterSerializer.descriptor)
+        })
+        element("Url", buildClassSerialDescriptor("Url") {
+            element("url", UrlSerializer.descriptor)
+            element("parameters", parameterSerializer.descriptor)
+        })
+    }
+
+    override fun deserialize(decoder: Decoder): RequestParametersFrom<T> {
+        require(decoder is JsonDecoder) // this class can be decoded only by Json
+
+        val element = decoder.decodeJsonElement()
+        return when {
+            "jsonString" in element.jsonObject -> RequestParametersFrom.Json(
+                decoder.json.decodeFromJsonElement<String>(element.jsonObject["jsonString"]!!),
+                decoder.json.decodeFromJsonElement(parameterSerializer, element.jsonObject["parameters"]!!)
+            )
+
+            "jwsSigned" in element.jsonObject -> run {
+                val parameters =
+                    decoder.json.decodeFromJsonElement(parameterSerializer, element.jsonObject["parameters"]!!)
+                val jwsSignedRaw = decoder.json.decodeFromJsonElement<String>(element.jsonObject["jwsSigned"]!!)
+                val jwsSignedFinal1 = JwsSigned.deserialize(jwsSignedRaw).getOrThrow()
+                val jws = JwsSigned<T>(
+                    jwsSignedFinal1.header,
+                    parameters,
+                    jwsSignedFinal1.signature,
+                    jwsSignedFinal1.plainSignatureInput
+                )
+                RequestParametersFrom.JwsSigned(
+                    jws,
```

Comment:
refactor to use only one line: `RequestParametersFrom.JwsSigned(jws, parameters)`